power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/common_solver_functions.hpp

```diff
@@ -57,17 +57,91 @@ template <symmetry_tag sym> inline void copy_y_bus(YBus<sym> const& y_bus, Compl
+        sources.begin(), sources.end(), ComplexValue<asymmetric_t>{}, std::plus{}, [&](Idx const source) {
+            ComplexValue<asymmetric_t> y_012;
+            y_012(0) = y_ref_012[source].y0;
+            y_012(1) = y_ref_012[source].y1;
+            y_012(2) = y_012(1); // y1 = y2
+            return y_012;
+        });
+    DoubleComplex const i_ref_1_t =
+        std::transform_reduce(sources.begin(), sources.end(), DoubleComplex{}, std::plus{},
+                              [&](Idx const source) { return input.source[source] * y_ref_012[source].y1; });
+    ComplexValue<asymmetric_t> const i_inj_t_012 = dot(get_sym_matrix_inv(), i_inj_t);
+
     for (Idx const source : sources) {
-        ComplexValue<sym> const u_ref{input.source[source]};
-        ComplexTensor<sym> const y_ref = y_bus.math_model_param().source_param[source];
-        output.source[source].i = dot(y_ref, u_ref - output.u[bus_number]);
+        DoubleComplex const y_ref_i_1_over_y_ref_t_1 = y_ref_012[source].y1 / y_ref_t_012[1];
+        DoubleComplex const i_inj_i_1_lhs =
+            y_ref_i_1_over_y_ref_t_1 * (input.source[source] * y_ref_t_012[1] - i_ref_1_t);
+        DoubleComplex const i_inj_i_0 = (y_ref_012[source].y0 / y_ref_t_012[0]) * i_inj_t_012(0);
+        DoubleComplex const i_inj_i_1 = i_inj_i_1_lhs + (y_ref_i_1_over_y_ref_t_1 * i_inj_t_012(1));
+        DoubleComplex const i_inj_i_2 = (y_ref_012[source].y1 / y_ref_t_012[2]) * i_inj_t_012(2);
+        ComplexValue<asymmetric_t> const i_inj_012{i_inj_i_0, i_inj_i_1, i_inj_i_2};
+        output.source[source].i = dot(get_sym_matrix(), i_inj_012);
         output.source[source].s = output.u[bus_number] * conj(output.source[source].i);
     }
 }
 
+// Current implementation avoids numerical instability. For details refer to:
+// https://github.com/PowerGridModel/power-grid-model-workshop/blob/experiment/source-calculation/source_calculation/source_calculation.ipynb
+template <symmetry_tag sym>
+inline void calculate_source_result(IdxRange const& sources, Idx const& bus_number, YBus<sym> const& y_bus,
+                                    PowerFlowInput<sym> const& input, SolverOutput<sym>& output,
+                                    IdxRange const& load_gens) {
+    if (sources.empty()) {
+        return;
+    }
+    ComplexValue<sym> const i_load_gen_bus =
+        std::transform_reduce(load_gens.begin(), load_gens.end(), ComplexValue<sym>{}, std::plus{},
+                              [&](Idx const load_gen) { return output.load_gen[load_gen].i; });
+    ComplexValue<sym> const i_inj_t = conj(output.bus_injection[bus_number] / output.u[bus_number]) - i_load_gen_bus;
+    if (sources.size() == 1) {
+        output.source[*sources.begin()].i = i_inj_t;
+        output.source[*sources.begin()].s = output.u[bus_number] * conj(output.source[*sources.begin()].i);
+    } else {
+        calculate_multiple_source_result(sources, y_bus, input, i_inj_t, output, bus_number);
+    }
```

Comment:
Out of curiosity: isn't single source just one case of multiple sources?

Reply:
Refer to this [notebook](https://github.com/PowerGridModel/power-grid-model-workshop/blob/experiment/source-calculation/source_calculation/source_calculation.ipynb) to see the mathematical difference in the calculation of one versus multiple sources. The detailed discussion we can have elsewhere, together with someone more experienced in power systems.